crates/astria-conductor/src/block_verifier.rs

```diff
@@ -0,0 +1,616 @@
+use ed25519_consensus::{
+    Signature,
+    VerificationKey,
+};
+use prost::Message;
+use tendermint::{
+    account::Id as AccountId,
+    block::{
+        parts,
+        CommitSig as TendermintCommitSig,
+        Height,
+        Id as BlockId,
+        Round,
+    },
+    chain::Id as ChainId,
+    crypto,
+    merkle,
+    vote::{
+        self,
+        CanonicalVote,
+    },
+    Hash,
+    Signature as TendermintSignature,
+    Time,
+};
+use tendermint_proto::types::CommitSig as RawCommitSig;
+use tracing::{
+    instrument,
+    warn,
+};
+
+use crate::tendermint::{
+    TendermintClient,
+    ValidatorSet,
+};
+
+const METRO_VALIDATOR_ADDRESS_PREFIX: &str = "metrovalcons";
+
+/// `BlockVerifier` is responsible for verifying the correctness of a block
+/// before executing it.
+/// It has two public functions: `validate_signed_namespace_data` and `validate_sequencer_block`.
+///
+/// `validate_signed_namespace_data` is used to validate the data received from the data
+/// availability layer. `validate_sequencer_block` is used to validate the blocks received from
+/// either the data availability layer or the gossip network.
+pub struct BlockVerifier {
+    tendermint_client: TendermintClient,
+}
+
+impl BlockVerifier {
+    pub fn new(tendermint_url: &str) -> eyre::Result<Self> {
+        Ok(Self {
+            tendermint_client: TendermintClient::new(tendermint_url.to_owned())
+                .wrap_err("failed to construct TendermintClient")?,
+        })
+    }
+
+    /// validates `SignedNamespaceData` received from Celestia.
+    /// This function verifies the block signature and checks that the data
+    /// was signed by the expected proposer for this block height.
+    pub(crate) async fn validate_signed_namespace_data(
+        &self,
+        data: &SignedNamespaceData<SequencerNamespaceData>,
+    ) -> eyre::Result<()> {
+        // verify the block signature
+        data.verify()
+            .wrap_err("failed to verify signature of signed namepsace data")?;
+
+        // get validator set for this height
+        let height: u64 = data
+            .data
+            .header
+            .height
+            .parse()
+            .wrap_err("failed to parse header height")?;
+        let validator_set = self
+            .tendermint_client
+            .get_validator_set(height - 1)
+            .await
+            .wrap_err("failed to get validator set")?;
+
+        // find proposer address for this height
+        let expected_proposer_address = validator_set
+            .get_proposer()
+            .wrap_err("failed to get proposer from validator set")?
+            .address;
+
+        // verify the namespace data signing public key matches the proposer address
+        let res_address = public_key_to_bech32_address(&data.public_key.0)
+            .wrap_err("failed to convert namespace data public key to address")?;
+        ensure!(
+            res_address == expected_proposer_address,
+            format!(
```

Comment:
This `format!` is IMO not necessary. I think `ensure!` would print the value of `res_address` and `expected_proposer_address`